src/lib/support/ScopedBuffer.h

```diff
@@ -0,0 +1,152 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      Defines scoped auto-free buffers for CHIP.
+ *
+ */
+
+#ifndef CHIP_SCOPED_BUFFER_H_
+#define CHIP_SCOPED_BUFFER_H_
+
+#include <lib/support/CHIPMem.h>
+
+namespace chip {
+namespace Platform {
+
+namespace Impl {
+
+/**
+ * Represents a memory buffer that is auto-freed in the destructor.
+ */
+template <class Impl>
+class ScopedMemoryBufferBase
+{
+public:
+    ScopedMemoryBufferBase() {}
+    ScopedMemoryBufferBase(const ScopedMemoryBufferBase &) = delete;
+    ScopedMemoryBufferBase & operator=(const ScopedMemoryBufferBase & other) = delete;
+
+    ~ScopedMemoryBufferBase() { Free(); }
+
+    void * Ptr() { return mBuffer; }
+    const void * Ptr() const { return mBuffer; }
+
+    template <typename T>
+    inline T * Ptr()
+    {
+        return static_cast<T *>(mBuffer);
+    }
+
+    template <typename T>
+    inline const T * Ptr() const
+    {
+        return static_cast<T *>(mBuffer);
+    }
+
+    ScopedMemoryBufferBase & Alloc(size_t size)
+    {
+        Free();
```

Comment:
Compilers manage to optimize out this null-check on the first `Alloc` call to make this 0-cost?  Nice.